agv_cv/noise_remover.cpp

```diff
@@ -0,0 +1,186 @@
+#include<stdio.h>
+#include<stdlib.h>
+#include"opencv/cv.h"
+#include<opencv2/highgui/highgui.hpp>
+#include<bits/stdc++.h>
+#include<typeinfo>
+#include<iostream>
+#include<vector>
+#include<time.h>
+#include<algorithm>
+#include<cmath>
+
+using namespace cv;
+using namespace std;
+
+struct co_ord_ {int x;int y;};
+typedef struct co_ord_ co_ord;
+
+void compute_quadratic(co_ord p1, co_ord p2, co_ord p3, float& a, float& b, float &c)
+{
+	a=( ((float)p1.x-p2.x)/(p1.y-p2.y)- ((float)p2.x-p3.x)/(p2.y-p3.y) )/((float)p1.y-p3.y);
+	b=( ((float)p1.x-p2.x)- a*(p1.y*p1.y- p2.y*p2.y) )/((float)p1.y-p2.y);
+	c=p1.x-a*p1.y*p1.y-b*p1.y;
+	return ;
+		
+}
+
+bool on_curve(co_ord p, float a, float b, float c) //checks if given point lies on curve
+{
+	float error_margin=25;
+	if(abs(a*p.y*p.y+b*p.y+c-p.x)<error_margin) return true;
+	else return false;
+}
+
+int main()
+{
+	Mat img =imread( "../images/noise.png", CV_LOAD_IMAGE_GRAYSCALE );
+	imshow("original", img);
+	waitKey(1);
+
+
+	/*int erosion_size = 3;  
+       	Mat element = getStructuringElement(cv::MORPH_RECT, cv::Size(2 * erosion_size + 1, 2 * erosion_size + 1), cv::Point(erosion_size, erosion_size) );
+	erode(img,img,element);
+	dilate(img,img,element);
+
+	imshow("eroded and diluted", img);
+	waitKey(3000);*/                              //erosion and dilution
+
+	medianBlur ( img, img, 15 );
+	//imshow("median blurred", img);
+	//waitKey(1);
+
+	Mat copy=img.clone();
+
+	vector<co_ord> points;
+	
+	//extract white pixels in vector points
+	int i, j, k;
+	for(i=0;i<img.rows;i++)
+		for(j=0;j<img.cols;j++)
+		{
+			if(img.at<uchar>(i,j)>=20) points.push_back({j,i});
+		}
+
+	float a, b, c;
+	float best_a=0, best_b=0, best_c=0;
+	int matches, best_matches=0;
+
+	//run ransac to fit a quadratic model
+	co_ord p1, p2, p3, best_p;
+	int iterations=100;
+	srand(12);
+	int tag;
+	while(iterations--)	
+	{
+		//randomly select three distinct points
+		tag=1;
+		do
+		{
+			p1=points[rand()%points.size()];
```

Comment:
Instead of choosing 3 points at random, decrease your domain so that the 3 points obtained always lie on the same line. In cases where there are a white patch in between the lanes, the algorithm may get confused as to which lane is what.